src/qibolab/_core/instruments/keysight/results.py

```diff
@@ -0,0 +1,76 @@
+"""Utils for result management."""
+
+import numpy as np
+from keysight import qcs
+
+from qibolab._core.execution_parameters import (
+    AcquisitionType,
+    AveragingMode,
+    ExecutionParameters,
+)
+from qibolab._core.sweeper import ParallelSweepers
+
+
+def fetch_result(
+    results: qcs.Results,
+    channel: qcs.Channels,
+    acquisition_type: AcquisitionType,
+    averaging: bool,
+) -> dict[qcs.Channels, np.ndarray]:
+    """Processes the QCS result object to return the appropiate results.
+
+    Arguments:
+        results (qcs.Results): QCS result object.
+        channel (qcs.Channels): Virtual acquisition channel to obtain results from.
+        acquisition_type (AcquisitionType): Acquisition type to be used.
+        averaging (bool): Flag for averaging results
+        sweeper_swaps_required (list[tuple[int, int]]): Array of axes pairs corresponding to swapped sweepers.
+
+    Returns:
+        raw (dict[qcs.Channels, np.ndarray]): Map of virtual channel to acquisition results.
+    """
+    if acquisition_type is AcquisitionType.RAW:
+        raw = results.get_trace(channel, averaging)
+    elif acquisition_type is AcquisitionType.INTEGRATION:
+        raw = results.get_iq(channel, averaging)
+    elif acquisition_type is AcquisitionType.DISCRIMINATION:
+        raw = results.get_classified(channel, averaging)
+    else:
+        raise ValueError("Acquisition type unrecognized")
+    return raw
+
+
+def parse_result(
+    result: np.ndarray, options: ExecutionParameters, sweepers: list[ParallelSweepers]
+) -> np.ndarray:
+    """Parses resulting numpy array into Qibolab expected array shape.
+
+    Arguments:
+        result (np.ndarray): Result array from QCS.
+        options (ExecutionParameters): Execution settings.
+        sweepers (list[ParallelSweepers]): Array of array of sweepers.
+
+    Returns:
+        parsed_result (np.ndarray): Parsed numpy array.
+    """
+    # For single shot, qibolab expects result format (nshots, ...)
+    # QCS returns (..., nshots), so we need to shuffle the arrays
+    if options.averaging_mode is AveragingMode.SINGLESHOT and len(sweepers) > 0:
+        tmp = np.zeros(options.results_shape(sweepers))
+        # For IQ data, QCS returns complex results
+        if options.acquisition_type is AcquisitionType.INTEGRATION:
+            for k in range(options.nshots):
+                tmp[k, ..., 0] = np.real(result[..., k])
+                tmp[k, ..., 1] = np.imag(result[..., k])
+        else:
+            for k in range(options.nshots):
+                tmp[k, ...] = result[..., k]
+
+    # For IQ data, QCS returns complex results
+    elif options.acquisition_type is AcquisitionType.INTEGRATION:
+        tmp = np.zeros(options.results_shape(sweepers))
+        tmp[..., 0] = np.real(result)
+        tmp[..., 1] = np.imag(result)
+    else:
+        return result
+    return tmp
```

Comment:
```suggestion
    if (
        options.averaging_mode is not AveragingMode.SINGLESHOT
        and options is not AcquisitionType.INTEGRATION
    ):
        return result

    # For single shot, qibolab expects result format (nshots, ...)
    # QCS returns (..., nshots), so we need to shuffle the arrays
    if options.averaging_mode is AveragingMode.SINGLESHOT:
        result = np.moveaxis(result, -1, 0)

    # For IQ data, QCS returns complex results
    if options.acquisition_type is not AcquisitionType.INTEGRATION:
        return result

    return np.moveaxis(np.stack([np.real(result), np.imag(result)]), 0, -1)
```

Ok, this is a proposal to simplify/deduplicate the current handling:
- if nothing needs to be done, early return
- you do not need to handle separately the case where sweepers are present or not (0 sweepers is a perfectly valid limit case)
- you can replace Python iteration by NumPy shape manipulations (that should even be virtual, and not affect the actual memory buffer, but only its view/access)
    - [`np.moveaxis`](https://numpy.org/doc/stable/reference/generated/numpy.moveaxis.html) is just doing the operation you wanted, bringing the innermost dimension to be the outermost
- similarly for stacking the complex components
    - `np.stack` is stacking the arrays in their outermost dimension
    - `np.moveaxis` is used to do the opposite operation of before: bring the outermost dimension to be the innermost

I hope it is actually more readable, but at least it avoids you to repeat the complex numbers handling for `AcquisitionType.INTEGRATION`, and it also saves the Python iteration, replacing it with a change of view, which should be quite more efficient (in case it will ever matter).

Of course, I did not test it, so that's up to you to accept it. In case there are any issue, but you like the overall proposal, I could help you fixing it.

Reply:
Yeah, I prefer NumPy in-memory reshaping over copying. Made a tiny tweak to make it work.

Changed in https://github.com/qiboteam/qibolab/pull/944/commits/5e490c14a0730d49d2748f6924d02d174621e19d